ports/x11/verbiste/verbiste-0.1.41/doc/french-conjugator.1
.\" $Id: french-conjugator.1.in,v 1.11 2013/07/18 03:13:05 sarrazip Exp $
.\" french-conjugator - Analyzer of conjugated French verbs
.\" verbiste - French conjugation system
.\" Copyright (C) 2003-2009 Pierre Sarrazin <http://sarrazip.com/>
.\"
.\" This program is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License
.\" as published by the Free Software Foundation; either version 2
.\" of the License, or (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
.\" 02110-1301, USA.
.\"
.\"
.TH french-conjugator "1" "April 6th, 2014" "" ""
.SH NAME
french-conjugator - conjugate French verbs
.SH SYNOPSIS
.B echo aimer | french-conjugator > result.txt
.SH DESCRIPTION
\fBfrench-conjugator\fR
reads the infinitive form of French verbs from
the command line or from standard input
and writes (to standard output) the complete conjugation of those
verbs, if they are known.
.PP
The standard input is not read if verbs are passed as command-line
arguments.
.PP
Each mode and tense is introduced by a line that starts with a hyphen
and a space, and ends with a colon.
The mode and tense in that line are always in English, regardless
of the user's current locale.
(This is meant to facilitate automatic parsing of the output.
For a French user interface, see the GNOME application and applet.)
The conjugation is ended with a line that only contains a hyphen.
If the given verb is unknown or not in the infinitive form,
only this line is written.
.PP
The command flushes its output buffer after finishing each answer.
This allows the command to be easily called from another program
through two pipes.
.PP
The command starts by loading its database from XML files
(stored typically in /usr/share/verbiste).
This takes some time, so it is a good idea to have the command
answer many requests instead of running it for each request.
.PP
The verbiste library's source archive contains Perl and Java
example programs that illustrate this technique.
.PP
There must not be any leading or trailing white spaces on the lines read
by the command.
.PP
In the \fBpast participle\fR tense, four lines are written: they correspond
in order
to the masculine singular, masculine plural, feminine singular and
feminine plural.
.SH OPTIONS
.TP
\fB\-\-help\fR
display a help page and exit
.TP
\fB\-\-version\fR
display version information and exit
.TP
\fB\-\-lang=L\fR
select the language to use (\fBfr\fR for French or \fBit\fR for Italian);
French is the default language
.TP
\fB\-\-mode=M\fR
only display mode \fBM\fR, where \fBM\fR can be
\fBinfinitive\fR,
\fBindicative\fR,
\fBconditional\fR,
\fBsubjunctive\fR,
\fBimperative\fR or
\fBparticiple\fR
.TP
\fB\-\-tense=T\fR
only display tense \fBT\fR, where \fBT\fR can be
\fBpresent\fR,
\fBpast\fR,
\fBimperfect\fR or
\fBfuture\fR
.TP
\fB\-\-pronouns\fR
show the pronouns
.TP
\fB\-\-all-infinitives\fR
print the infinitive form of all the verbs in the knowledge base,
one per line, unsorted; other command-line arguments are ignored
.SH EXAMPLES
$ french-conjugator aimer
.br
- infinitive present:
.br
aimer
.br
- indicative present:
.br
aime
.br
aimes
.br
aime
.br
aimons
.br
aimez
.br
aiment
.br
- indicative imperfect:
.br
aimais
.br
aimais
.br
aimait
.br
[...]
.br
- participle past:
.br
aim\('e
.br
aim\('es
.br
aim\('ee
.br
aim\('ees
.br
-
.br
.SH LICENSE
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.
.SH AUTHOR
See the verbiste(3) manual page.
.SH BUGS
See the verbiste(3) manual page.
.SH SEE ALSO
verbiste(3), french-deconjugator(1).


ports/x11/verbiste/verbiste-0.1.41/src/commands/french-conjugator.cpp
/*  $Id: french-conjugator.cpp,v 1.31 2012/11/18 19:56:04 sarrazip Exp $
    french-conjugator.cpp - Conjugation of French verbs

    verbiste - French conjugation system
    Copyright (C) 2003-2010 Pierre Sarrazin <http://sarrazip.com/>

    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License
    as published by the Free Software Foundation; either version 2
    of the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
    02110-1301, USA.
*/


#include "Command.h"

#include "gui/conjugation.h"

#ifdef ENABLE_NLS
#ifdef HAVE_GETOPT_LONG
#include <unistd.h>
#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif
#include <getopt.h>
#endif
#endif  /* ENABLE_NLS */

#include <langinfo.h>

#include <iostream>
#include <vector>
#include <string.h>

using namespace std;
using namespace verbiste;


static const char *commandName = "french-conjugator";


#if defined(ENABLE_NLS) && defined(HAVE_GETOPT_LONG)
static struct option knownOptions[] =
{
    { "help",           no_argument,            NULL, 'h' },
    { "version",        no_argument,            NULL, 'v' },
    { "lang",           required_argument,      NULL, 'l' },
    { "mode",           required_argument,      NULL, 'm' },
    { "tense",          required_argument,      NULL, 't' },
    { "template",       required_argument,      NULL, 'e' },
    { "pronouns",       no_argument,            NULL, 'p' },
    { "all-infinitives",no_argument,            NULL, 'i' },

    { NULL, 0, NULL, 0 }  // marks the end
};
#endif


class ConjugatorCommand : public Command
{
public:

    ConjugatorCommand(const string &conjugationFilename,
                      const string &verbsFilename,
                      FrenchVerbDictionary::Language lang) throw(logic_error)
      : Command(conjugationFilename, verbsFilename, lang),
        reqMode(INVALID_MODE),
        reqTense(INVALID_TENSE),
        reqTemplate(),
        includePronouns(false),
        aspirateH(false)
    {
    }

    virtual ~ConjugatorCommand()
    {
    }

    // Constraints:
    Mode reqMode;
    Tense reqTense;
    string reqTemplate;
    bool includePronouns;

protected:

    virtual void processInputWord(const std::string &inputWord);

private:

    bool aspirateH;

    void displayTense(FrenchVerbDictionary &fvd,
                        const string &radical,
                        const TemplateSpec &templ,
                        Mode mode,
                        Tense tense);

    void displayConjugation(FrenchVerbDictionary &fvd,
                        const string &infinitive,
                        const string &tname,
                        const TemplateSpec &templ);
};


void
ConjugatorCommand::displayTense(FrenchVerbDictionary &fvd,
                                const string &radical,
                                const TemplateSpec &templ,
                                Mode mode,
                                Tense tense)
{
    if (reqMode != INVALID_MODE && mode != reqMode)
        return;
    if (reqTense != INVALID_TENSE && tense != reqTense)
        return;

    cout << "- " << FrenchVerbDictionary::getModeName(mode)
                << " " << FrenchVerbDictionary::getTenseName(tense) << ":\n";


    typedef vector<string> VS;
    typedef vector<VS> VVS;

    VVS conjug;
    fvd.generateTense(radical, templ, mode, tense, conjug,
                      includePronouns, aspirateH, lang == FrenchVerbDictionary::ITALIAN);

    for (VVS::const_iterator p = conjug.begin(); p != conjug.end(); p++)
    {
        for (VS::const_iterator i = p->begin(); i != p->end(); i++)
        {
            if (i != p->begin())
                cout << ", ";
            string s = *i;

            cout << s;
        }
        cout << "\n";
    }
}


void
ConjugatorCommand::displayConjugation(FrenchVerbDictionary &fvd,
                                        const string &infinitive,
                                        const string &tname,
                                        const TemplateSpec &templ)
{
    try
    {
        aspirateH = fvd.isVerbStartingWithAspirateH(infinitive);

        string radical = FrenchVerbDictionary::getRadical(infinitive, tname);

        if (lang == FrenchVerbDictionary::FRENCH || lang == FrenchVerbDictionary::ITALIAN)
        {
            displayTense(fvd, radical, templ, INFINITIVE_MODE, PRESENT_TENSE);

            displayTense(fvd, radical, templ, INDICATIVE_MODE, PRESENT_TENSE);
            displayTense(fvd, radical, templ, INDICATIVE_MODE, IMPERFECT_TENSE);
            displayTense(fvd, radical, templ, INDICATIVE_MODE, FUTURE_TENSE);
            displayTense(fvd, radical, templ, INDICATIVE_MODE, PAST_TENSE);

            displayTense(fvd, radical, templ, CONDITIONAL_MODE, PRESENT_TENSE);

            displayTense(fvd, radical, templ, SUBJUNCTIVE_MODE, PRESENT_TENSE);
            displayTense(fvd, radical, templ, SUBJUNCTIVE_MODE, IMPERFECT_TENSE);

            displayTense(fvd, radical, templ, IMPERATIVE_MODE, PRESENT_TENSE);

            displayTense(fvd, radical, templ, PARTICIPLE_MODE, PRESENT_TENSE);
            displayTense(fvd, radical, templ, PARTICIPLE_MODE, PAST_TENSE);
        }
        if (lang == FrenchVerbDictionary::ITALIAN)
            displayTense(fvd, radical, templ, GERUND_MODE, PRESENT_TENSE);

        if (lang == FrenchVerbDictionary::GREEK)
        {
            displayTense(fvd, radical, templ, PRESENT_INDICATIVE, ACTIVE_TENSE);
            displayTense(fvd, radical, templ, PRESENT_INDICATIVE, PASSIVE_TENSE);
            displayTense(fvd, radical, templ, PRESENT_SUBJUNCTIVE, ACTIVE_TENSE);
            displayTense(fvd, radical, templ, PRESENT_SUBJUNCTIVE, PASSIVE_TENSE);
            displayTense(fvd, radical, templ, PRESENT_IMPERATIVE, IMPERATIVE_ACTIVE_TENSE);
            displayTense(fvd, radical, templ, PRESENT_IMPERATIVE, IMPERATIVE_PASSIVE_TENSE);
            displayTense(fvd, radical, templ, PRESENT_GERUND, PRESENT_TENSE);
            displayTense(fvd, radical, templ, PAST_IMPERFECT_INDICATIVE, ACTIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_IMPERFECT_INDICATIVE, PASSIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_PERFECT_INDICATIVE, ACTIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_PERFECT_INDICATIVE, PASSIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_PERFECT_SUBJUNCTIVE, ACTIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_PERFECT_SUBJUNCTIVE, PASSIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_PERFECT_IMPERATIVE, IMPERATIVE_ACTIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_PERFECT_IMPERATIVE, IMPERATIVE_PASSIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_PERFECT_INFINITIVE, PAST_PERFECT);
        }
    }
    catch (logic_error &e)
    {
        return;
    }
}


/*virtual*/
void
ConjugatorCommand::processInputWord(const string &inputWord)
{
    const char *tname = (reqTemplate.empty() ? NULL : reqTemplate.c_str());
    const std::set<std::string> *templateSet = NULL;
    std::set<std::string> singleton;

    if (tname == NULL)  // if no specific template requested
    {
        // Use the template associated with the verb, if known:
        templateSet = &fvd->getVerbTemplateSet(inputWord);
    }
    else
    {
        singleton.insert(tname);
        templateSet = &singleton;
    }

    for (std::set<std::string>::const_iterator it = templateSet->begin();
                                               it != templateSet->end(); ++it)
    {
        if (it != templateSet->begin())
            cout << "-\n";  // separate conjugations

        const string &tname = *it;
        const TemplateSpec *templ = fvd->getTemplate(tname);

        if (templ != NULL)
            displayConjugation(*fvd, inputWord, tname, *templ);
    }

    cout << "-" << endl;  // marks the end of the answer, and flushes it
}


#if defined(ENABLE_NLS) && defined(HAVE_GETOPT_LONG)

static
void
displayVersionNo()
{
    cout << commandName << ' ' << VERSION << '\n';
}


static
void
displayHelp()
{
    cout << '\n';

    displayVersionNo();

    cout << "Part of " << PACKAGE << " " << VERSION << "\n";

    cout <<
"\n"
"Copyright (C) " COPYRIGHT_YEARS " Pierre Sarrazin <http://sarrazip.com/>\n"
"This program is free software; you may redistribute it under the terms of\n"
"the GNU General Public License.  This program has absolutely no warranty.\n"
    ;

    cout <<
"\n"
"Options:\n"
"--help             Display this help page and exit\n"
"--version          Display this program's version number and exit\n"
"--lang=L           Select language L (fr for French, it for Italian)\n"
"                   Default is French.\n"
"--mode=M           Only display mode M (infinitive, indicative, etc)\n"
"--tense=T          Only display tense T (present, past, etc)\n"
"--template=T       Use template T to conjugate the verbs\n"
"--pronouns         Include pronouns in the displayed conjugation\n"
"--all-infinitives  Print the names of all known verbs, one per line (unsorted)\n"
"\n"
"See the " << commandName << "(1) manual page for details.\n"
"\n"
    ;
}

#endif


int
main(int argc, char *argv[])
{
    setlocale(LC_CTYPE, "");

    Mode reqMode = INVALID_MODE;
    Tense reqTense = INVALID_TENSE;
    string reqTemplate;
    bool includePronouns = false;
    bool listAllInfinitives = false;
    string langCode = "fr";

    #if defined(ENABLE_NLS) && defined(HAVE_GETOPT_LONG)

    /*  Interpret the command-line options:
    */
    int c;
    try
    {
        do
        {
            c = getopt_long(argc, argv, "hvm:t:l:p", knownOptions, NULL);

            switch (c)
            {
                case EOF:
                    break;  // nothing to do

                case 'l':
                    langCode = optarg;
                    if (FrenchVerbDictionary::parseLanguageCode(langCode) == FrenchVerbDictionary::NO_LANGUAGE)
                    {
                        cerr << commandName << ": invalid language code " << optarg << "\n";
                        return EXIT_FAILURE;
                    }
                    break;

                case 'm':
                    reqMode = FrenchVerbDictionary::convertModeName(optarg);
                    if (reqMode == INVALID_MODE)
                    {
                        cerr << commandName << ": invalid mode " << optarg << "\n";
                        return EXIT_FAILURE;
                    }
                    break;

                case 't':
                    reqTense = FrenchVerbDictionary::convertTenseName(optarg);
                    if (reqTense == INVALID_TENSE)
                    {
                        cerr << commandName << ": invalid tense " << optarg << "\n";
                        return EXIT_FAILURE;
                    }
                    break;

                case 'e':
                    reqTemplate = optarg;
                    break;

                case 'p':
                    includePronouns = true;
                    break;

                case 'v':
                    displayVersionNo();
                    return EXIT_SUCCESS;

                case 'h':
                    displayHelp();
                    return EXIT_SUCCESS;

                case 'i':
                    listAllInfinitives = true;
                    break;

                default:
                    displayHelp();
                    return EXIT_FAILURE;
            }
        } while (c != EOF && c != '?');
    }
    catch (const exception &e)
    {
        cerr << commandName << ": exception: " << e.what() << endl;
        return EXIT_FAILURE;
    }

    #else

    int optind = 1;

    #endif

    try
    {
        FrenchVerbDictionary::Language lang = FrenchVerbDictionary::parseLanguageCode(langCode);
        if (lang == FrenchVerbDictionary::NO_LANGUAGE)
            lang = FrenchVerbDictionary::FRENCH;

        string conjFN, verbsFN;
        FrenchVerbDictionary::getXMLFilenames(conjFN, verbsFN, lang);

        ConjugatorCommand cmd(conjFN, verbsFN, lang);

        if (listAllInfinitives)
            return cmd.listAllInfinitives(cout);

        cmd.reqMode = reqMode;
        cmd.reqTense = reqTense;
        cmd.reqTemplate = reqTemplate;
        cmd.includePronouns = includePronouns;

        if (!reqTemplate.empty() &&
                    cmd.getFrenchVerbDictionary().getTemplate(reqTemplate) == NULL)
        {
            cerr << commandName << ": invalid conjugation template "
                                                << optarg << "\n";
            return EXIT_FAILURE;
        }

        return cmd.run(argc - optind, argv + optind);
    }
    catch (const exception &e)
    {
        cerr << commandName << ": exception: " << e.what() << endl;
        return EXIT_FAILURE;
    }
}
